### PR TITLE
Improve Reading Content from WattTime

### PR DESCRIPTION
**Pull request by juzuluag**
Issue Number: (585)

## Summary
- Fetch data from WattTime service using streams instead of strings
- Cache in memory a BalanceAuthority instance to avoid hitting WattTime service all the time for the same item.

## Changes

- Improve retrieving the data when the header is available, then client can start to stream it, instead of waiting until all the content (string) is in available in memory.
- Change how WattTime Client reads content from the requests and to cache BA.

## Checklist

- [x] Local Tests Passing?
- [x] CICD and Pipeline Tests Passing?
- [x] Documentation?

## Are there API Changes?
N/A
## Is this a breaking change?
N/A
## Anything else?
Attached are screenshots about before and after the changes about the number of requests made to WattTime for a batch workload for the same location.
### Before 
Screen Shot 2022-08-12 at 3 28 55 PM

### After
Screen Shot 2022-08-12 at 11 53 09 PM

### CODE REVIEW

File: src/CarbonAware.Tools/CarbonAware.Tools.WattTimeClient/src/Configuration/WattTimeClientConfiguration.cs [hunk 1]
```diff
@@ -23,6 +23,12 @@ public class WattTimeClientConfiguration
     /// </summary>
     public string BaseUrl { get; set; } = "https://api2.watttime.org/v2/";
 
+    /// <summary>
+    /// Gets or sets the cached expiration time (in seconds) for a BalancingAuthority instance.
+    /// It defaults to 60 secs.
+    /// </summary>
+    public int BACacheExpirationInSecs { get; set; } = 60;
```
Comment (gfmatthews) [inline src/CarbonAware.Tools/CarbonAware.Tools.WattTimeClient/src/Configuration/WattTimeClientConfiguration.cs]: After our chat, I was thinking about this, how about 5 minutes to correspond with the resolution that WattTime gives forecasts?
```suggestion
    public int BACacheExpirationInSecs { get; set; } = 300;
```

File: src/CarbonAware.Tools/CarbonAware.Tools.WattTimeClient/src/WattTimeClient.cs [hunk 2]
```diff
@@ -61,9 +65,10 @@ public async Task<IEnumerable<GridEmissionDataPoint>> GetDataAsync(string balanc
             { QueryStrings.BalancingAuthorityAbbreviation, balancingAuthorityAbbreviation }
         };
 
-        var result = await this.MakeRequestAsync(Paths.Data, parameters, tags);
-
-        return JsonSerializer.Deserialize<List<GridEmissionDataPoint>>(result, options) ?? new List<GridEmissionDataPoint>();
+        using (var result = await this.MakeRequestGetStreamAsync(Paths.Data, parameters, tags))
+        {
+            return await JsonSerializer.DeserializeAsync<List<GridEmissionDataPoint>>(result, options) ?? throw new WattTimeClientException($"Error getting forecasts for {balancingAuthorityAbbreviation}");
```
Comment (gfmatthews) [inline src/CarbonAware.Tools/CarbonAware.Tools.WattTimeClient/src/WattTimeClient.cs:70]: Thank you for making this error more clear

File: src/CarbonAware.Tools/CarbonAware.Tools.WattTimeClient/src/WattTimeClient.cs [hunk 3]
```diff
@@ -325,4 +305,29 @@ private string BuildUrlWithQueryString(string url, IDictionary<string, string> q
 
         return result;
     }
+
+    private async Task<BalancingAuthority> GetBalancingAuthorityFromCacheAsync(string latitude, string longitude)
+    {
+        var key = new Tuple<string, string>( latitude, longitude );
+        var balancingAuthority = await this.memoryCache.GetOrCreateAsync(key, async entry =>
```
Comment (gfmatthews) [inline src/CarbonAware.Tools/CarbonAware.Tools.WattTimeClient/src/WattTimeClient.cs:312]: Does this have a chance to throw anything if something is up with the cache?

**Review by gfmatthews — APPROVED** [review]
Nothing blocking. Great work making this improvement. One question - should we add a test in the test suite for checking these cases where we're pulling the BA from the cache and also not from the cache?

File: src/CarbonAware.Tools/CarbonAware.Tools.WattTimeClient/src/WattTimeClient.cs (hunk 3 above)
Comment (juzuluag) [reply src/CarbonAware.Tools/CarbonAware.Tools.WattTimeClient/src/WattTimeClient.cs:312]: Based on MS doc it doesn't. Worst case is that catch can grow until resources are consumed. Having the expiration time helps to control to some degree that.

File: src/CarbonAware.Tools/CarbonAware.Tools.WattTimeClient/src/Configuration/WattTimeClientConfiguration.cs [hunk 4]
```diff
@@ -23,6 +23,12 @@ public class WattTimeClientConfiguration
     /// </summary>
     public string BaseUrl { get; set; } = "https://api2.watttime.org/v2/";
 
+    /// <summary>
+    /// Gets or sets the cached expiration time (in seconds) for a BalancingAuthority instance.
+    /// It defaults to 300 secs.
+    /// </summary>
+    public int BACacheExpirationInSecs { get; set; } = 300;
```
Comment (bderusha) [inline src/CarbonAware.Tools/CarbonAware.Tools.WattTimeClient/src/Configuration/WattTimeClientConfiguration.cs]: The WattTime docs recommend, `querying /ba-from-loc at least once a month with coordinates to determine a device's BA region.` 

We can probably use a more aggressive default value.  I would say at least an hour (3600), but I would be comfortable going as high as 1 day (86400).  What do you think?

File: GettingStarted.md [hunk 5]
```diff
@@ -157,6 +157,12 @@ You can configure the verbosity of the application error messages by setting the
 CarbonAwareVars__VerboseApi="true"
 ```
 
+### WattTimeClient Caching BalancingAuthority
+To improve the interaction with WattTime service, the client added a setting named `BACacheExpirationInSecs` to control when a `BalancingAuthority` instance will expire from caching it in memory. This is useful for batch requests, where multiple WattTime requests can be done for the same location, improving the overall communication with the service. Default value is `300 seconds`.
+```bash
+WattTimeClient__BACacheExpirationInSecs="90"
```
Comment (bderusha) [inline GettingStarted.md]:
```suggestion
WattTimeClient__BalancingAuthorityCacheTTL="90"
```

NIT:  prefer common web cache terminology: `TTL` (time to live)
prefer to use full name `BalancingAuthority` for consistency of naming within the project,

File: GettingStarted.md [hunk 6]
```diff
@@ -157,6 +157,12 @@ You can configure the verbosity of the application error messages by setting the
 CarbonAwareVars__VerboseApi="true"
 ```
 
+### WattTimeClient Caching BalancingAuthority
+To improve the interaction with WattTime service, the client added a setting named `BACacheExpirationInSecs` to control when a `BalancingAuthority` instance will expire from caching it in memory. This is useful for batch requests, where multiple WattTime requests can be done for the same location, improving the overall communication with the service. Default value is `300 seconds`.
```
Comment (bderusha) [inline GettingStarted.md]:
```suggestion
To improve performance communicating with the WattTime API service, the client caches the data mapping location coordinates to balancing authorities.  By default, this data is stored in an in-memory cache for `XXXX` seconds, but expiration can be configured using the setting `BalancingAuthorityCacheTTL` (Set to "0" to not use cache).  The regional boundaries of a balancing authority tend to be stable, but as they can change, the [WattTime documentation](https://www.watttime.org/api-documentation/#determine-grid-region) recommends not caching for longer than 1 month.
```

Suggested more declarative language and referenced the docs.

**Review by bderusha — APPROVED** [review]
Code looks great.  Just some comments about naming and defaults.

File: src/CarbonAware.Tools/CarbonAware.Tools.WattTimeClient/src/Configuration/WattTimeClientConfiguration.cs (hunk 4 above)
Comment (juzuluag) [reply src/CarbonAware.Tools/CarbonAware.Tools.WattTimeClient/src/Configuration/WattTimeClientConfiguration.cs]: @bderusha , no preference tbh. I'll default it to 1 day then.

**Pull request merged**